book/python/api.md

```diff
@@ -32,111 +30,163 @@ The recommended way to create your topology structure is by using the [Applicati
 
-The `wallaroo` module provides an [ApplicationBuilder](#wallarooapplicationbuilder) that facilitates the creation of this data structure. When `ApplicationBuilder` is used, a topology can be built using its methods and its structure can be returned by calling `ApplicationBuilder.build()`.
+#### Partial Pipelines
 
-For a simple application with a decoder, computation, and encoder, this function may look like
+A pipeline includes one or more sources. You use `wallaroo.source(...)` to define a stream originating from a source. Each source stream can be followed by one or more computation stages. A linear sequence from a source through zero or more computations constitutes a partial pipeline. For example:
 
 ```python
-def application_setup(args):
-    in_host, in_port = wallaroo.tcp_parse_input_addrs(args)[0]
-    out_host, out_port = wallaroo.tcp_parse_output_addrs(args)[0]
-
-    ab = wallaroo.ApplicationBuilder("My Application")
-    ab.new_pipeline("pipeline 1",
-                    wallaroo.TCPSourceConfig(in_host, in_port, decoder))
-    ab.to(computation)
-    ab.to_sink(wallaroo.TCPSinkConfig(out_host, out_port, encoder))
-    return ab.build()
+inputs = wallaroo.source("Source Name", source_config)
+partial_pipeline = inputs.to(my_computation)
 ```
 
-#### `args`
+This defines a partial pipeline that could be diagrammed as followed:
 
-Since the application is run in an embedded Python runtime, it does not have standard access to `sys.argv` from which various options parsers could be used. Instead, Wallaroo provides `application_setup` with `args`: a list of the string command-line arguments it has received.
+```
+Source -> my_computation ->
+```
 
-### ApplicationBuilder
+The hanging arrow at the end indicates that the pipeline is partial. We can still add more stages, and to complete the pipeline we need one or more sinks.
+You create complete pipeline by terminating a partial pipeline with a call to `to_sink` or `to_sinks`. For example:
 
-The `ApplicationBuilder` class in `wallaroo` is a utility for constructing application topologies. It also provides some additional validation on the input parameters.
+```python
+inputs = wallaroo.source("Source Name", source_config)
+complete_pipeline = (inputs
+    .to(my_computation)
+    .to_sink(sink_config))
+```
 
-#### `ApplicationBuilder` Methods
+Our pipeline is now complete:
 
-##### `__init__(name)`
+```
+Source -> my_computation -> Sink
+```
 
-Create a new application with the name `name`.
+Unless a call to `to` using a stateless computation is preceded by a call to `key_by` (which partitions messages by key), there are no guarantees around the order messages will be processed in. That's because Wallaroo might parallelize a stateless computation if that is beneficial for scaling. That means the execution graph for the above pipeline could look like this:
```

Comment:
"around the order in which messages will be processed"

Reply:
done